plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizer.java

```diff
@@ -0,0 +1,827 @@
+        }
+        String body = response.body();
+        try {
+            return json.readValue(body, cls);
+        }
+        catch (Exception e) {
+            throw new OpaQueryException.DeserializeFailed(e);
+        }
+    }
+
+    protected boolean queryOpa(OpaQueryInput input)
+    {
+        OpaQueryResult result = tryGetResponseFromOpa(input, opaPolicyUri, OpaQueryResult.class);
+        if (result.result == null) {
+            return false;
+        }
+        return result.result;
+    }
+
+    protected boolean queryOpaWithSimpleAction(SystemSecurityContext context, String operation)
+    {
+        OpaQueryInputAction action = new OpaQueryInputAction.Builder().operation(operation).build();
+        OpaQueryInput input = new OpaQueryInput(context, action);
+        return queryOpa(input);
+    }
+
+    protected boolean queryOpaWithSimpleResource(SystemSecurityContext context, String operation, OpaQueryInputResource resource)
+    {
+        OpaQueryInputAction action = new OpaQueryInputAction.Builder().operation(operation).resource(resource).build();
+        OpaQueryInput input = new OpaQueryInput(context, action);
+        return queryOpa(input);
+    }
+
+    @Override
+    public void checkCanImpersonateUser(SystemSecurityContext context, String userName)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource.Builder().user(new OpaQueryInputResource.User(userName)).build();
+
+        if (!queryOpaWithSimpleResource(context, "ImpersonateUser", resource)) {
+            SystemAccessControl.super.checkCanImpersonateUser(context, userName);
+        }
+    }
+
+    @Override
+    public void checkCanSetUser(Optional<Principal> principal, String userName)
+    {
+        // This method is deprecated and is called for any identity, let's no-op
+    }
+
+    @Override
+    public void checkCanExecuteQuery(SystemSecurityContext context)
+    {
+        if (!queryOpaWithSimpleAction(context, "ExecuteQuery")) {
+            SystemAccessControl.super.checkCanExecuteQuery(context);
+        }
+    }
+
+    @Override
+    public void checkCanViewQueryOwnedBy(SystemSecurityContext context, Identity queryOwner)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource.Builder().user(new OpaQueryInputResource.User(queryOwner)).build();
+        if (!queryOpaWithSimpleResource(context, "ViewQueryOwnedBy", resource)) {
+            SystemAccessControl.super.checkCanViewQueryOwnedBy(context, queryOwner);
+        }
+    }
+
+    @Override
+    public Collection<Identity> filterViewQueryOwnedBy(SystemSecurityContext context, Collection<Identity> queryOwners)
+    {
+        return queryOwners.parallelStream().filter(queryOwner -> queryOpaWithSimpleResource(context, "FilterViewQueryOwnedBy", new OpaQueryInputResource.Builder().user(new OpaQueryInputResource.User(queryOwner)).build())).collect(Collectors.toSet());
```

Comment:
Interesting, we took this idea from the original implementation by @sbernauer @nightkr @maltesander @lfrancke @fhennig , as well as the Ranger implementation by @dprophet on #13297

That's not to say this is the right approach of course :)
The main complexity from OPA is that the answer to an authorization query may not be straightforward - the concept of "has any access" is quite vague in OPA because policies are code (rego) and not a static policy document like what you'd see on Ranger or the File Based Access Control.

Whoever writes the rego policy may choose to create a rule to allow access to `FilterTables` based on whether the user can `SelectFromColumns`, but we cannot (IMHO) make that decision for them.

Perhaps `Filter*` is not the right operation name for these cases, but I believe we still need to call the operation something unique (e.g. `HasAnyAccess`) so as to allow policy developers to decide what "having any access" entails.

Unless policies are extremely complex, it should be straightforward to map `HasAnyAccess` or `Filter*` to specific operations within OPA/rego. @sbernauer has suggested that we should perhaps create a set of small rego helpers to make writing policies easier for most users (e.g. by mapping `Filter*` operations to something else) while keeping the specificity in what the plugin sends so users can choose to be very specific if they so desire.

Reply:
Like I said, I don't know what makes sense for OPA. But whatever it is, it would be best if it helped the implementor in getting the semantics right with minimal effort. Perhaps it can be done generically without writing any additional code in OPA, perhaps not. I'm not going to tell what it should be, though.